link-grammar/prepare/exprune.c

```diff
@@ -429,25 +377,33 @@ void expression_prune(Sentence sent, Parse_Options opts)
 		for (w = 0; w < sent->length; w++)
 		{
 			/* For every expression in word */
-			for (x = sent->word[w].x; x != NULL; x = x->next)
-			{
-				DBG(pass, w, "l->r pass before marking");
-				N_deleted += mark_dead_connectors(ctxt.ct, w, x->exp, '-');
-				DBG(pass, w, "l->r pass after marking");
-			}
-			for (x = sent->word[w].x; x != NULL; x = x->next)
+			for (X_node **xp = &sent->word[w].x; *xp != NULL; /* See: NEXT */)
 			{
+				X_node *x = *xp;
+
 				DBG(pass, w, "l->r pass before purging");
-				x->exp = purge_Exp(x->exp);
+				x->exp = purge_Exp(ctxt.ct, w, x->exp, '-', &N_deleted);
 				DBG(pass, w, "l->r pass after purging");
+
+				/* Get rid of X_nodes with NULL exp */
+				if (x->exp == NULL)
+				{
+					*xp = x->next; /* NEXT - set current X_node to the next one */
+					x->next = free_later;
+					free_later = x;
```

Comment:
The idea was to not pollute the instruction cache with the `free()` code, and later let the `free()` loop to contain mainly the `free()` code. This seemed to gain some marginal speedup.

What may gain a significant speedup, on the expense of some memory, is not to free anything at all until `sentence_delete()`. This may "waste" several MB of memory for long sentences, but such sentences need anyway many hundreds of MB in the parsing stage so this will still have a relatively small effect on the total memory consumption. Also, in that case the pool allocator can be used for Exp/Exp_list, for an additional speedup.

Reply:
> What may gain a significant speedup

I see that it is not so - there are not so many empty X_nodes to make a significant speedup in this manner. (But not freeing stuff in `Exp_purge()` may worth trying.)